Race condition in 0.9

I will try to provide a nicer MWE later, but I ran into a problem while working on some proprietary data. The file in question parses without problem in CSV 0.8.3, but fails under the same options in 0.9. Adding `ntasks=1` in CSV 0.9 solves the issue, which makes me suspect a race condition.

cc: @dmbates

Filename passed as a string; kwargs: 
```julia
comment="#", 
delim=';', 
missingstring=["NA",""],
downcast=true,
drop=[1,2],
decimal=','
```

Error + stacktrace:
```julia
TaskFailedException

nested task error: AssertionError: task_col.column isa Vector{UInt32}

Stacktrace:

[1] syncrefs!(#unused#::Type{String}, col::CSV.Column, task_col::CSV.Column, task_rows::Int64)

@ CSV ~/.julia/packages/CSV/eB98W/src/file.jl:474

[2] multithreadpostparse(ctx::CSV.Context, ntasks::Int64, pertaskcolumns::Vector{Vector{CSV.Column}}, rows::Vector{Int64}, rowchunkguess::Int64, finalrows::Int64, j::Int64, col::CSV.Column)

@ CSV ~/.julia/packages/CSV/eB98W/src/file.jl:425

[3] (::CSV.var"#26#31"{CSV.Context, Vector{Int64}, Vector{Vector{CSV.Column}}, Int64, Int64, CSV.Column, Int64, Int64})()

@ CSV ./threadingconstructs.jl:169

...and 1 more exception.

    sync_end(::Channel{Any})@task.jl:369
    macro expansion@task.jl:388[inlined]
    CSV.File(::CSV.Context, ::Bool)@file.jl:257
    File@file.jl:223[inlined]
    #File#23@file.jl:219[inlined]
    top-level scope@Local: 2[inlined]
```

--- a/src/context.jl
+++ b/src/context.jl
@@ -23,6 +23,7 @@ Fields:
   * `userprovidedtype`: whether the column type was provided by the user or not; this affects whether we'll promote a column's type while parsing, or emit a warning/error depending on `strict` keyword arg
   * `willdrop`: whether we'll drop this column from the final columnset; computed from select/drop keyword arguments; this will result in a column type of `HardMissing` while parsing, where an efficient parser is used to "skip" a field w/o allocating any parsed value
   * `pool`: computed from `pool` keyword argument; `true` is `1.0`, `false` is `0.0`, everything else is `Float64(pool)`; once computed, this field isn't mutated at all while parsing; it's used in type detection to determine whether a column will be pooled or not once a type is detected; 
+  * `columnspecificpool`: if `pool` was provided via Vector or Dict by user, then `true`, other `false`; if `false`, then only string column types will attempt pooling
   * `column`: the actual column vector to hold parsed values; field is typed as `AbstractVector` and while parsing, we do switches on `col.type` to assert the column type to make code concretely typed
   * `refpool`: if the column is pooled (or might be pooled in single-threaded case), this is the column-specific `RefPool` used to track unique parsed values and their `UInt32` ref codes
   * `lock`: in multithreaded parsing, we have a top-level set of `Vector{Column}`, then each threaded parsing task makes its own copy to parse its own chunk; when synchronizing column types/pooled refs, the task-local `Column` will `lock(col.lock)` to make changes to the parent `Column`; each task-local `Column` shares the same `lock` of the top-level `Column`
@@ -36,6 +37,7 @@ mutable struct Column
     userprovidedtype::Bool
     willdrop::Bool
     pool::Float64
+    columnspecificpool::Bool
     # lazily/manually initialized fields
     column::AbstractVector
     refpool::RefPool
@@ -45,8 +47,8 @@ mutable struct Column
     endposition::Int
     options::Parsers.Options
 
-    Column(type::Type, anymissing::Bool, userprovidedtype::Bool, willdrop::Bool, pool::Float64) =
-        new(type, anymissing, userprovidedtype, willdrop, pool)
+    Column(type::Type, anymissing::Bool, userprovidedtype::Bool, willdrop::Bool, pool::Float64, columnspecificpool::Bool) =
+        new(type, anymissing, userprovidedtype, willdrop, pool, columnspecificpool)
 end
 
 function Column(type::Type, options::Parsers.Options=nothing)
@@ -54,7 +56,7 @@ function Column(type::Type, options::Parsers.Options=nothing)
     col = Column(type === Missing ? HardMissing : T,
         type >: Missing,
         type !== NeedsTypeDetection,
-        false, NaN)
+        false, NaN, false)
     if options !== nothing
         col.options = options
     end
@@ -64,7 +66,7 @@ end
 # creating a per-task column from top-level column
 function Column(x::Column)
     @assert isdefined(x, :lock)
-    y = Column(x.type, x.anymissing, x.userprovidedtype, x.willdrop, x.pool)
+    y = Column(x.type, x.anymissing, x.userprovidedtype, x.willdrop, x.pool, x.columnspecificpool)
     y.lock = x.lock # parent and child columns _share_ the same lock
     if isdefined(x, :options)
         y.options = x.options
@@ -483,15 +485,22 @@ end
         if pool isa AbstractVector
             length(pool) == ncols || throw(ArgumentError("provided `pool::AbstractVector` keyword argument doesn't match detected # of columns: `$(length(pool)) != $ncols`"))
             for i = 1:ncols
-                columns[i].pool = getpool(pool[i])
+                col = columns[i]
+                col.pool = getpool(pool[i])
+                col.columnspecificpool = true
             end
         elseif pool isa AbstractDict
             for i = 1:ncols
-                columns[i].pool = getpool(getordefault(pool, names[i], i, 0.0))
+                col = columns[i]
+                col.pool = getpool(getordefault(pool, names[i], i, 0.0))
+                col.columnspecificpool = true
             end
             checkinvalidcolumns(pool, "pool", ncols, names)
         else
             finalpool = getpool(pool)
+            for col in columns
+                col.pool = finalpool
+            end
         end
     end
 
@@ -595,12 +604,6 @@ end
             rowsguess = ceil(Int64, ((len - datapos) / weightedavgbytesperrow) * 1.01)
             debug && println("single-threaded estimated rows = $origrowsguess, multi-threaded estimated rows = $rowsguess")
             debug && println("multi-threaded column types sampled as: $columns")
-            # check if we need to adjust column pooling
-            if finalpool == 0.0 || finalpool == 1.0
-                for col in columns
-                    col.pool = finalpool
-                end
-            end
         else
             debug && println("something went wrong chunking up a file for multithreaded parsing, falling back to single-threaded parsing")
             threaded = false

--- a/src/detection.jl
+++ b/src/detection.jl
@@ -336,7 +336,7 @@ ColumnProperties(T) = ColumnProperties(T, 0x00)
     end
 end
 
-function findchunkrowstart(ranges, i, buf, opts, downcast, ncols, rows_to_check, columns, columnlock, @nospecialize(stringtype), totalbytes, totalrows, succeeded)
+function findchunkrowstart(ranges, i, buf, opts, downcast, ncols, rows_to_check, columns, origcoltypes, columnlock, @nospecialize(stringtype), totalbytes, totalrows, succeeded)
     pos = ranges[i]
     len = ranges[i + 1]
     while pos <= len
@@ -359,9 +359,11 @@ function findchunkrowstart(ranges, i, buf, opts, downcast, ncols, rows_to_check,
         parsedncols = rowsparsed = 0
         columnprops = Vector{ColumnProperties}(undef, ncols)
         for i = 1:ncols
-            if columns[i].type === NeedsTypeDetection
-                columnprops[i] = ColumnProperties(typecode(columns[i].type))
+            if origcoltypes[i] === NeedsTypeDetection
+                columnprops[i] = ColumnProperties(NEEDSTYPEDETECTION)
             else
+                # if column type is already set, that means user provided manually
+                # so we don't care about sampling that column's values for type
                 columnprops[i] = ColumnProperties(0x00)
             end
         end
@@ -461,9 +463,10 @@ function findrowstarts!(buf, opts, ranges, ncols, columns, @nospecialize(stringt
     succeeded = Threads.Atomic{Bool}(true)
     N = length(ranges) - 2
     lock = ReentrantLock()
+    origcoltypes = Type[col.type for col in columns]
     @sync for i = 2:(length(ranges) - 1)
         Threads.@spawn begin
-            findchunkrowstart(ranges, i, buf, opts, downcast, ncols, rows_to_check, columns, lock, stringtype, totalbytes, totalrows, succeeded)
+            findchunkrowstart(ranges, i, buf, opts, downcast, ncols, rows_to_check, columns, origcoltypes, lock, stringtype, totalbytes, totalrows, succeeded)
         end
     end
     return totalbytes[] / totalrows[], succeeded[]

--- a/src/file.jl
+++ b/src/file.jl
@@ -269,7 +269,9 @@ function File(ctx::Context, @nospecialize(chunking::Bool=false))
         # cleanup our columns if needed
         for col in columns
 @label processcolumn
-            if !isdefined(col, :refpool) && pooled(col)
+            if col.type === NeedsTypeDetection && pooled(col)
+                # we get here if user specified pool=true for at least this column
+                # but we never encountered any non-missing values
                 col.type = Missing
                 col.column = allocate(Pooled, finalrows)
                 col.refpool = RefPool(Missing)
@@ -279,12 +281,11 @@ function File(ctx::Context, @nospecialize(chunking::Bool=false))
                 col.column = MissingVector(finalrows)
                 col.pool = 0.0
             end
-            if pooled(col)
-                makepooled!(col)
-            elseif col.column isa Vector{UInt32}
-                # check if final column should be PooledArray or not
-                poolval = !isnan(col.pool) ? col.pool : !isnan(ctx.pool) ? ctx.pool : DEFAULT_POOL
-                if ((length(col.refpool.refs) - 1) / finalrows) <= poolval
+            if col.column isa Vector{UInt32}
+                if pooled(col)
+                    makepooled!(col)
+                elseif ((length(col.refpool.refs) - 1) / finalrows) <= col.pool
+                    # check if final column should be PooledArray or not
                     makepooled!(col)
                 else
                     # cardinality too high, so unpool
@@ -394,7 +395,7 @@ function multithreadpostparse(ctx, ntasks, pertaskcolumns, rows, rowchunkguess, 
         elseif T === Float64 && T2 <: Integer
             # one chunk parsed as Int, another as Float64, promote to Float64
             ctx.debug && println("multithreaded promoting column $j to float")
-            if pooled(col)
+            if task_col.column isa Vector{UInt32}
                 task_col.refpool.refs = convert(Refs{Float64}, task_col.refpool.refs)
             else
                 task_col.column = convert(SentinelVector{Float64}, task_col.column)
@@ -409,17 +410,13 @@ function multithreadpostparse(ctx, ntasks, pertaskcolumns, rows, rowchunkguess, 
             # one chunk parsed all missing values, but another chunk had a typed value, promote to that
             # while keeping all values `missing` (allocate by default ensures columns have all missing values)
             ctx.debug && println("multithreaded promoting column $j from missing on task $i")
-            task_col.column = allocate(pooled(col) ? Pooled : T, task_rows)
+            task_col.column = allocate(maybepooled(col) && (T isa StringTypes || col.columnspecificpool) ? Pooled : T, task_rows)
         end
         # synchronize refs if needed
-        if pooled(col) || (isnan(col.pool) && col.type isa StringTypes)
+        if maybepooled(col) && (T isa StringTypes || col.columnspecificpool)
             if !isdefined(col, :refpool) && isdefined(task_col, :refpool)
                 # this case only occurs if user explicitly passed pool=true
                 # but we only parsed `missing` values during sampling
-                col.refpool = task_col.refpool
-            elseif isdefined(col, :refpool) && pooltype(col) !== T && isdefined(task_col, :refpool)
-                # we pooled/detected one type while sampling, but parsing promoted to another type
-                # if the task_col has a refpool, then we know it's the promoted type
                 col.refpool = task_col.refpool
             elseif isdefined(task_col, :refpool)
                 syncrefs!(col.type, col, task_col, task_rows)
@@ -427,14 +424,14 @@ function multithreadpostparse(ctx, ntasks, pertaskcolumns, rows, rowchunkguess, 
         end
     end
 @label threadedprocesscolumn
-    if pooled(col)
-        makechain!(col.type, pertaskcolumns, col, j, ntasks)
-        # pooled columns are the one case where we invert the order of arrays;
-        # i.e. we return PooledArray{T, ChainedVector{T}} instead of ChainedVector{T, PooledArray{T}}
-        makepooled!(col)
-    elseif isnan(col.pool) && col.type isa StringTypes
-        poolval = !isnan(col.pool) ? col.pool : !isnan(ctx.pool) ? ctx.pool : DEFAULT_POOL
-        if ((length(col.refpool.refs) - 1) / finalrows) <= poolval
+    # @show maybepooled(col), col.type, isdefined(col, :column)
+    if maybepooled(col) && (col.type isa StringTypes || col.columnspecificpool)
+        if pooled(col)
+            makechain!(col.type, pertaskcolumns, col, j, ntasks)
+            # pooled columns are the one case where we invert the order of arrays;
+            # i.e. we return PooledArray{T, ChainedVector{T}} instead of ChainedVector{T, PooledArray{T}}
+            makepooled!(col)
+        elseif ((length(col.refpool.refs) - 1) / finalrows) <= col.pool
             col.pool = 1.0
             @goto threadedprocesscolumn
         else
@@ -519,7 +516,7 @@ function makechain!(::Type{T}, pertaskcolumns, col, j, ntasks) where {T}
             col.column = ChainedVector([convert(Vector{Bool}, pertaskcolumns[i][j].column::vectype(T)) for i = 1:ntasks])
         elseif col.type !== Union{} && col.type <: SmallIntegers
             col.column = ChainedVector([convert(Vector{col.type}, pertaskcolumns[i][j].column::vectype(T)) for i = 1:ntasks])
-        elseif pooled(col)
+        elseif maybepooled(col) && (col.type isa StringTypes || col.columnspecificpool)
             col.column = ChainedVector([pertaskcolumns[i][j].column::Vector{UInt32} for i = 1:ntasks])
         else
             col.column = ChainedVector([parent(pertaskcolumns[i][j].column) for i = 1:ntasks])
@@ -783,7 +780,7 @@ function detectcell(buf, pos, len, row, rowoffset, i, col, ctx, rowsguess)::Tupl
     end
 @label done
     # if we're here, that means we found a non-missing value, so we need to update column
-    if pooled(col) || maybepooled(col) || (isnan(col.pool) && newT isa StringTypes)
+    if maybepooled(col) && (newT isa StringTypes || col.columnspecificpool)
         column = allocate(Pooled, rowsguess)
         col.refpool = RefPool(newT)
         val = getref!(col.refpool, newT, val isa PosLen ? PosLenString(buf, val, opts.e) : val)
@@ -955,7 +952,7 @@ end
 @noinline function promotetostring!(ctx::Context, buf, pos, len, rowsguess, rowoffset, columns, ::Type{customtypes}, column_to_promote, numwarnings, limit, stringtype) where {customtypes}
     cols = [i == column_to_promote ? columns[i] : Column(Missing, columns[i].options) for i = 1:length(columns)]
     col = cols[column_to_promote]
-    if pooled(col) || maybepooled(col) || isnan(col.pool)
+    if maybepooled(col)
         col.refpool = RefPool(stringtype)
         col.column = allocate(Pooled, rowsguess)
     else

--- a/src/utils.jl
+++ b/src/utils.jl
@@ -19,7 +19,7 @@ finaltype(::Type{NeedsTypeDetection}) = Missing
 coltype(col) = ifelse(col.anymissing, Union{finaltype(col.type), Missing}, finaltype(col.type))
 
 pooled(col) = col.pool == 1.0
-maybepooled(col) = 0.0 < col.pool < 1.0
+maybepooled(col) = 0.0 < col.pool
 
 function getpool(x::Real)::Float64
     if x isa Bool
@@ -101,7 +101,7 @@ function allocate!(columns, rowsguess)
         # if the type hasn't been detected yet, then column will get allocated
         # in the detect method while parsing
         if col.type !== NeedsTypeDetection
-            if pooled(col) || maybepooled(col) || (isnan(col.pool) && col.type isa StringTypes)
+            if maybepooled(col) && (col.type isa StringTypes || col.columnspecificpool)
                 col.column = allocate(Pooled, rowsguess)
                 if !isdefined(col, :refpool)
                     col.refpool = RefPool(col.type)

